### PR TITLE
Conditional deploy of CRDs on shoot

### PR DESCRIPTION
**Pull request by MartinWeindel**
**How to categorize this PR?**

/area control-plane
/kind technical-debt

**What this PR does / why we need it**:
For shooted seeds some CRDs of the dns-controller-manager are deployed by two managed resources. The managed resource for deploying the shoot-dns-extension itself (from the ControllerDeployment) and the managed resource for the extension object (from the upstream seed/soil) currently contain these CRDs.
This PR removes the second one ("extension-shoot-dns-service-shoot-keep") and the dns-controller-manager running in the control plane deploys these CRDs, but only if the shoot is no seed.

Additionally the CRD for unused and obsolete `DNSLocks` is removed.

**Which issue(s) this PR fixes**:
Fixes #

**Special notes for your reviewer**:

**Release note**:

```other operator
Conditional deploy of CRDs on shoot.
```

### CODE REVIEW

File: pkg/controller/lifecycle/actuator.go [hunk 1]
```diff
@@ -758,44 +756,8 @@ func (a *actuator) cleanupShootDNSEntries(helper *common.ShootDNSEntriesHelper)
...
-		labels := c.GetLabels()
-		if labels == nil {
-			labels = map[string]string{}
-		}
-		labels[v1beta1constants.ShootNoCleanup] = "true"
-		c.SetLabels(labels)
-	}
-
-	if err = managedresources.CreateFromUnstructured(ctx, a.Client(), namespace, KeptShootResourcesName, false, "", crds, true, nil); err != nil {
+	// TODO (Martin Weindel) delete obsolete managed resource, the following 3 lines can be deleted after release 0.55.0
```
[the remark this answers is not in the document]
Comment (MartinWeindel) [reply pkg/controller/lifecycle/actuator.go]: Addressed by deleting all managed resources named `extension-shoot-dns-service-shoot-keep` on startup. As pointed out by Tim, it is save to remove the finalizers and delete these MRs, as they have `.spec.keepObjects=true` anyway.

File: cmd/gardener-extension-shoot-dns-service/app/app.go [hunk 2]
```diff
@@ -106,9 +110,64 @@ func (o *Options) run(ctx context.Context) error {
...
+func (m *migrations) AddToManager(_ context.Context, mgr manager.Manager) error {
+	m.client = mgr.GetClient()
+	m.log = mgr.GetLogger().WithName("migrations")
+	return mgr.Add(m)
+}
+
+func (m *migrations) deleteObsoleteManagedResources(ctx context.Context) error {
+	log := m.log.WithName("deleteObsoleteManagedResources")
+
+	list := &metav1.PartialObjectMetadataList{}
```
Comment (timuthy) [inline cmd/gardener-extension-shoot-dns-service/app/app.go:150]: As the controller already works with a `ManagedResource` cache, you could also directly use `&ManagedResourceList{}`, I guess.

File: cmd/gardener-extension-shoot-dns-service/app/app.go [hunk 3]
```diff
@@ -106,9 +110,64 @@ func (o *Options) run(ctx context.Context) error {
...
+	list.SetGroupVersionKind(resourcesv1alpha1.SchemeGroupVersion.WithKind("ManagedResourceList"))
+	if err := m.client.List(ctx, list); err != nil {
```
Comment (timuthy) [inline cmd/gardener-extension-shoot-dns-service/app/app.go:152]: Does it make sense to use `.metadata.name` field selector here?

File: cmd/gardener-extension-shoot-dns-service/app/app.go (hunk 2 above)
Comment (MartinWeindel) [reply cmd/gardener-extension-shoot-dns-service/app/app.go]: changed

File: cmd/gardener-extension-shoot-dns-service/app/app.go (hunk 3 above)
Comment (MartinWeindel) [reply cmd/gardener-extension-shoot-dns-service/app/app.go]: Using `.metadata.name` field selector requires to add a field indexer for the cache with lengthy additional code. I don't think it is worth the effort as it runs only once on startup.

**Comment by MartinWeindel** [comment]
rebased to resolve merge conflict

**Review by timuthy — APPROVED** [review]
/lgtm

**Pull request merged**